### PR TITLE
gRPC-based KMS plugin service

### LINKED ISSUE #51965: API for Key Management Service (KMS) plugins for secrets encryption key management
/kind feature

In 1.7 we added the capability to encrypt secrets at the database layer using keys provided in a file on disk on the master. The next logical step is to protect those keys using a Key Management Service such as Vault or the KMS's provided by the major cloud providers. 

In 1.8 via [48574](https://github.com/kubernetes/kubernetes/pull/48574) we added a plugin mechanism and an implementation for Google Cloud KMS.  But this is based on the existing cloud provider integration point, so doesn't help the in-flight [Vault implementation](https://github.com/kubernetes/kubernetes/pull/51760) or any other non-cloud providers. See [design doc for more details](https://docs.google.com/document/d/19SbqLrSC_MpmxbLCxasubRJLaZxgqfw1KeIwg3fHHLU/edit).

As @luxas states in the PR:

> For beta, this has to work with out of tree cloud providers, and API server <-> KMS integration should > be done via some kind of interface, but not by compiling the KMS code into the API server.

This bug will track the effort to add that API interface.

@lavalamp 

@kubernetes/sig-api-machinery-feature-requests
@kubernetes/sig-auth-misc

### PR DESCRIPTION
**Pull request by wu-qiang**
**What this PR does / why we need it**:
Implement for issue https://github.com/kubernetes/kubernetes/issues/51965
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #51965

**Special notes for your reviewer**:
@destijl @sakshamsharma @deads2k @ericchiang
The implementation based on the document https://docs.google.com/document/d/1S_Wgn-psI0Z7SYGvp-83ePte5oUNMr4244uanGLYUmw/edit
**Release note**:

```release-note
Implement envelope service with gRPC, so that KMS providers can be pulled out from API server.
```

### CODE REVIEW

**Comment by ericchiang** [comment]
/ok-to-test
cc @kubernetes/sig-api-machinery-pr-reviews 
cc @kubernetes/sig-auth-pr-reviews

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 1]
```diff
@@ -150,18 +150,37 @@ func GetPrefixTransformers(config *ResourceConfig) ([]value.PrefixTransformer, e
...
+				envelopeService, pluginFound, err = KMSPluginRegistry.getPlugin(provider.KMS.Name, f)
+				if err != nil {
+					return nil, fmt.Errorf("could not configure KMS plugin %q, %v", provider.KMS.Name, err)
+				}
+				if pluginFound == false {
+					return nil, fmt.Errorf("KMS plugin %q not found", provider.KMS.Name)
+				}
+			} else {
+				// Get gRPC client service with remote config
+				envelopeService, err = envelope.NewEnvelopeService(
```
Comment (sakshamsharma) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: Rename to `grpcClientService,err := envelope.NewGRPCService(...)`? (Refer to comment in grpc_service.go)

Comment (sakshamsharma) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: Rename to `grpcClientService,err := envelope.NewGRPCService(...)`? (Refer to comment in grpc_service.go)

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 2]
```diff
@@ -150,18 +150,37 @@ func GetPrefixTransformers(config *ResourceConfig) ([]value.PrefixTransformer, e
...
-			if err != nil {
-				return nil, fmt.Errorf("could not configure KMS plugin %q, %v", provider.KMS.Name, err)
-			}
-			if pluginFound == false {
-				return nil, fmt.Errorf("KMS plugin %q not found", provider.KMS.Name)
+
+			var envelopeService envelope.Service
+			remoteConfig := provider.KMS.RemoteServer
+			if remoteConfig == nil {
+				// There should be no KMS provider plugins on API server side in future.
```
Comment (sakshamsharma) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: I agree with this comment. Should we remove this block altogether? @destijl

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+)
+
+type gRPCService struct {
+	// gRPC client instance
+	kmsClient  KmsServiceClient
+	connection *grpc.ClientConn
+}
+
+// NewEnvelopeService returns an envelope.Service which use gRPC to communicate the remote KMS provider.
+func NewEnvelopeService(endpoint, serverCert, clientCert, clientKey string) (Service, error) {
```
Comment (sakshamsharma) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Perhaps renaming this to `NewGRPCService` or `NewRemoteKMSService` might be more informative. The current name sounds a bit too generic IMO.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/service.pb.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+var _ context.Context
+var _ grpc.ClientConn
+
+// This is a compile-time assertion to ensure that this generated file
+// is compatible with the grpc package it is being compiled against.
+const _ = grpc.SupportPackageIsVersion4
+
+// Client API for KmsService service
+
+type KmsServiceClient interface {
```
Comment (sakshamsharma) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/service.pb.go]: The Go  standard dictates camel case of the form: `KMSServiceClient`.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	protocol, addr, err := parseEndpoint(endpoint)
```
Comment (sakshamsharma) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Endpoint might not be provided in the configuration. `url.Parse` does not give error on an empty input string. You may want to test for it inside `parseEndpoint` and return a custom error.

**Review by sakshamsharma — COMMENTED** [review]
First pass review comments. Also, could you please rename the PR to something like "gRPC-based KMS plugin service"? :smile:

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 3 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Okay, will modify it to "NewGRPCService".

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 5 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Now the error for empty endpoint will be handled in below switch default block. Of course, test endpoint explicitly is more readable. I will add it.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/service.pb.go (hunk 4 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/service.pb.go]: Will update the proto file to fix it.

**Comment by jbeda** [comment]
Talked about this in SIG-Auth and wanted to draw attention to https://github.com/kubernetes/kubernetes/issues/55729.  This is another type of extension mechanism where the API server is calling out.  I'd love to figure out how this relates to other types of webhook/callout mechanisms.

(I also understand that this is a bit of a different beast in that it is lower level than many of the other web hooks and may have extra performance requirements.)

**Comment by kksriram** [comment]
@jbeda I hear you about introducing yet another type of extension mechanism. I am looking at  @kubernetes/sig-api-machinery-pr-reviews for input so we can make corrections sooner than later. Making changes to the approach now seems the right time as 1.9 enters slush/freeze and we can get changes ready in time for _possible_ inclusion in 1.10 as alpha.

**Comment by lavalamp** [comment]
This is not targeted for 1.9, right?

**Comment by kksriram** [comment]
@lavalamp much as I would like to, I don't think so. Idk who makes the call on these, though.

**Comment by kksriram** [comment]
@lavalamp NVM, I see Anthony's post [here](https://groups.google.com/forum/#!topic/kubernetes-sig-release/SUJ8C0BJ_j0) on sig-release. I guess its up to sig-auth to target it for 1.9.

**Comment by dims** [comment]
Let's see what they say ... @kubernetes/sig-auth-pr-reviews @kubernetes/sig-auth-api-reviews @kubernetes/sig-auth-feature-requests Do we do this in 1.9 or move it to 1.10?

**Comment by smarterclayton** [comment]
Discussion in sig-arch:

* We need to get more formal about what types of APIs are ok
* There was a good analogy about two types of apis
  * declarative, user facing - follows our core API conventions
  * imperative, downward facing, non-declarative - typically gRPC or other
domain appropriate - i.e. etcd, CRI, CNI, CSI

An example was made that webhook style APIs, while sometimes imperative,
can also be user facing.  Provisionally the use of gRPC here was warranted
but a longer discussion is needed and to finalize those decisions and
record them.

**Comment by deads2k** [comment]
> Let's see what they say ... @kubernetes/sig-auth-pr-reviews @kubernetes/sig-auth-api-reviews @kubernetes/sig-auth-feature-requests Do we do this in 1.9 or move it to 1.10?

This is not targeted at 1.9.

**Comment by vineet-garg** [comment]
@jbeda @smarterclayton Is there a decision to go ahead with gRPC for this PR ?

**Comment by destijl** [comment]
This is 1.10 but we'd also like to get the implementation of the KMSs that *use* this API into 1.10. Which means we want this in early in the 1.10 cycle. It's especially important since by imposing the requirement for this API we removed one complete implementation and blocked two more that were ready to merge.

While I appreciate there might be a larger API design discussion going on I'd like to see some priority on making a decision about this particular API. Can we get a "yes go ahead with gRPC" or "no do it this way" in the week after Kubecon?

@lavalamp would it be worth discussion on this topic at the API update or deep dive at Kubecon?

Looking at the calendar seems like the next api SIG meeting probably isn't until 20 Dec.

**Comment by wu-qiang** [comment]
This pull request depends on https://github.com/kubernetes/kubernetes/pull/55184
After 55184 merged, I will update code base on its modification.

**Comment by luxas** [comment]
cc @kubernetes/sig-cluster-lifecycle-pr-reviews

**Comment by khenidak** [comment]
I have few comments, some addressable in the PR others will require a larger discussion at the architecture level. I have split them over 2 comments.

**Call for a unified api-server identity**
1. Api server *calls-out* to many logically *external entities* for various *functions* such as KMS, various types of web hooks (admission as an example), and many others will be added. I understand that they might be running on the same cluster but this is not relevant to this context.
2. Each *external entity* requires a configuration that consists of many things one of which is **identity** used by the api-server.
3. Each entity/function is configured and spec-ed independently. Asking the user to configure an object/yaml file with some client-cert-bundle or a like. 
4. This situation leads to an operational nightmare, specifically in identity/key rotation. Worst case a  scenario a compromised api-server (or etcd) will require the operator to disable *unknown number* compromised identities depending on configuration. please keep in mind **there is no easy way currently to enumerate all the identities used by api-server**

** Proposal **

1. Introduce a new *identity* object, that has a *type* (such as: cert/token/username+pwd) and references a *secret* which carries the actual data. Any *call-out* configuration will explicitly reference the identity object.
   
2. introduce a new *call-out* object that wraps the call out configuration details (such as: endpoint metadata). That *call-out* object *must* use the identity object.  An implementation such as this PR should only be concerned with its own configuration not identity/nor endpoint details. 

Both will make it easier to 1) rotate the identity 2) enumerate the entities/functions when needed. The change can be gradually introduced over multiple major/minor releases.

**Comment by khenidak** [comment]
The only concern i have is: Users can not run the grpc server as deployment on the same cluster (Not even using static manifests). This grpc server has to run outside the cluster using systemd (configuring restarts etc..) or the likes. This means 

1. Existing bootstrap-ing tools will have to be modified. 
2. There is no way we can look into this executable using existing clusters monitoring/logging tools.
3. Key rotation will force users manually configure each api-server.

**Comment by kksriram** [comment]
@khenidak Your [comment](https://github.com/kubernetes/kubernetes/pull/55684#issuecomment-350503493) about server identity to identify the api-server outside the cluster is a necessary component to this solution. This came up at the [Identity BoF](http://sched.co/D32z) and the [Container Identity WG Update](http://sched.co/CszN) at KubeCon. @destijl I don't have the slides you walked through - but I thought this was an item on the roadmap for 1.10?

**Comment by ericchiang** [comment]
> This situation leads to an operational nightmare, specifically in identity/key rotation. Worst case a scenario a compromised api-server (or etcd) will require the operator to disable unknown number compromised identities depending on configuration. please keep in mind there is no easy way currently to enumerate all the identities used by api-server

I think this comes down to "give your webhooks (audit, authentication, authorization, kms) all the same credentials"

You can do this today by pointing all of the webhook configs at common, on disk credentials. Later you might be able to expand the rotation strategies with https://github.com/kubernetes/kubernetes/issues/55968, or we could introduce watchers and say "if this credential file ever changes, reload the credentials"

While this is related, I don't think its a blocker for this PR, and should be discussed in a different issue.

**Comment by ericchiang** [comment]
> The only concern i have is: Users can not run the grpc server as deployment on the same cluster (Not even using static manifests).

Why wouldn't you be able to run this as static manifests?

**Comment by khenidak** [comment]
@ericchiang re-same identity while giving all webhook the same credentials will work, its not exactly an acceptable solution for regulated environments, where identities are compartmentalized - as a best practice - to limit the blast radius when one is compromised.

As for running the gRpc as static manifest/pod. I retract. Yes i think it will work.

**Comment by ericchiang** [comment]
The use of gRPC still needs to be discussed in sig-architecture (see https://github.com/kubernetes/kubernetes/pull/55684#issuecomment-345136687).

https://github.com/kubernetes/community/tree/master/sig-architecture

@liggitt volunteered to bring this up for their meeting tomorrow. If that meeting's skipped it due to KubeCon/the holidays, we'll have to wait until people are back in January.

**Comment by lavalamp** [comment]
I don't think it is a good precedent for sig architecture to review things at this level of detail unless sigs are not managing to come to an agreement among themselves. I don't think anyone actually objects to the gRPC mechanism? I recommend saving the red tape for cases where it's actually useful.

**Comment by bgrant0607** [comment]
@lavalamp The reason this was brought up in SIG Arch. was the issue of consistency: what types of extension API mechanisms should we use? What amount of consistency do we need/want? Should it be decided on a per-component basis? How extensive do the conventions need to be? There at least need to be meta-requirements, such as (presumably) every extension mechanism must have a way to version it. At the moment, we have execution-based plugins (kubelet, kubectl), webhooks (apiserver, scheduler), and grpc (kubelet).

**Comment by bgrant0607** [comment]
@khenidak @kksriram Please start an email thread on the identity topic and include at least the identity WG, SIG Auth, SIG API Machinery, and SIG Cluster Lifecycle.

**Comment by lavalamp** [comment]
@bgrant0607 I don't disagree that sig architecture should be worried about consistency. I just think it should wait for someone actually doing it wrong to intervene. There were zero actual objections to this mechanism. Intervening on everything in this class of change (perhaps I draw it more broadly than you would) will give the sig too much to look at and not add value to the rest of the project.

**Comment by liggitt** [comment]
Outcome of discussion this morning:
* gRPC use here is appropriate
* should be scoped to local access via socket, using file permissions for access, not adding multi-user/authn/authz layers now
* the gRPC interface should consider how it will ensure the called service is version-compatible if a new field that modifies the semantics is added. As an example, the kubelet's CRI gRPC interface has a Version method that allows introspecting a CRI implementation. Related issues:
  * https://github.com/kubernetes/kubernetes/issues/28789
  * https://github.com/kubernetes/kubernetes/issues/28642
  * https://github.com/kubernetes/kubernetes/blob/e87d8511b1ed35c1dd9d5ffe521b6cb4098523dd/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L16-L19

**Comment by wu-qiang** [comment]
@liggitt Updated the design document as comments.
1) Remove TCP and TLS support.
2) Add a new "Version" method in interface.

**Comment by wu-qiang** [comment]
@liggitt @sakshamsharma  Do below 3 modifications:
1. Remove TCP and TLS support.
2. Add "Version" method in gRPC interface.
3. Merge with the modification in https://github.com/kubernetes/kubernetes/pull/55184

**Comment by wu-qiang** [comment]
The pull request of the design document https://github.com/kubernetes/community/pull/1581
It's based on the content and comments in https://docs.google.com/document/d/1S_Wgn-psI0Z7SYGvp-83ePte5oUNMr4244uanGLYUmw/edit

**Comment by destijl** [comment]
Agree with divorcing this from solving the whole identity problem and just going with a local socket for now. Thanks for coming to an agreement on gRPC. I think the initial objections are addressed, lets get this reviewed and move forward: @lavalamp @deads2k @cheftako

**Comment by destijl** [comment]
@immutablet please review and test this version with the Google KMS

**Comment by kksriram** [comment]
@lavalamp @deads2k @cheftako 
Bump. 
Any comments on the code? I see tests passed. Would like to have this merged so we can progress on other related topics (identity, testing with cloud providers etc.)

**Comment by ericchiang** [comment]
cc @gyuho on gRPC API versioning.

**Comment by gyuho** [comment]
Just sharing how etcd handles API versioning with gRPC, if anybody is interested.

We do not expose any version API via gRPC. Instead we have stream/unary RPC interceptor to gRPC server. For instance,

```go
grpc.NewServer(
   grpc.UnaryInterceptor(newUnaryInterceptor(EtcdServer)),
   grpc.StreamInterceptor(newStreamInterceptor(EtcdServer)),
)
```

So long as the version information is available somewhere in runtime, interceptor checks version compatibility on every gRPC call to filter out incompatible API calls.

/cc @xiang90

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+// The gRPC implementation for envelope.Service.
+type gRPCService struct {
+	// gRPC client instance
+	kmsClient  kmsapi.KMSServiceClient
+	connection *grpc.ClientConn
+}
+
+// NewGRPCService returns an envelope.Service which use gRPC to communicate the remote KMS provider.
+func NewGRPCService(endpoint string) (Service, error) {
+	glog.Infof("Configure KMS provider with endpoint: %s", endpoint)
```
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Would it be better to set this at something like glog.V(4).Infof ?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+	response, err := kmsClient.Version(context.Background(), request)
+	if err != nil {
+		return fmt.Errorf("failed get version from remote KMS provider: %v", err)
+	}
+	if response.Version != kmsapiVersion {
+		return fmt.Errorf("KMS provider api version %s is not supported, only %s is supported now",
+			response.Version, kmsapiVersion)
+	}
+
+	glog.Infof("KMS provider %s initialized, version: %s", response.RuntimeName, response.RuntimeVersion)
```
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Again maybe V4?

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go [hunk 8]
```diff
@@ -83,4 +83,6 @@ type KMSConfig struct {
 	CacheSize int `json:"cachesize,omitempty"`
 	// configfile is the path to the configuration file for the named KMS provider.
 	ConfigFile string `json:"configfile"`
+	// the gRPC server listening address, for example "unix:///var/run/kms-provider.sock".
+	Endpoint string `json:"endpoint"`
```
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go]: For backward compatibility shouldn't this be both "// +optional" and omitempty?

Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go]: Yes, it's better to set them optional.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 6 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Okay.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go [hunk 9]
```diff
@@ -82,5 +82,9 @@ type KMSConfig struct {
 	// +optional
 	CacheSize int `json:"cachesize,omitempty"`
 	// configfile is the path to the configuration file for the named KMS provider.
-	ConfigFile string `json:"configfile"`
+	// +optional
+	ConfigFile string `json:"configfile,omitempty"`
+	// the gRPC server listening address, for example "unix:///var/run/kms-provider.sock".
+	// +optional
+	Endpoint string `json:"endpoint,omitempty"`
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go]: At this point you need validation on the type.  Not all combinations are valid and it needs to be enforced.

Also, I don't see a version on this config. Without a version, we won't be able to easily have structural change.  For instance, later allowing a remote listening address will require more fields that should probably be logically grouped.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+	protocol, addr, err := parseEndpoint(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	dialer := func(addr string, timeout time.Duration) (net.Conn, error) {
+		return net.DialTimeout(protocol, addr, timeout)
+	}
+
+	connection, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithDialer(dialer))
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: `WithDialer` is required?  It reads weird to me that we have to provide our own.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+	if u.Scheme != unixProtocol {
+		return "", "", fmt.Errorf("unsupported scheme %q for remote KMS provider", u.Scheme)
+	}
+	return unixProtocol, u.Path, nil
+}
+
+// Check the KMS provider API version.
+// Only matching kubeRuntimeAPIVersion is supported now.
+func checkAPIVersion(kmsClient kmsapi.KMSServiceClient) error {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Mock this out demonstrating what you'd do on a version negotiation.  I can see that the information is available, but this should be returning an interface used in the rest of the code and instantiating the impl to be the v1beta1 impl.  

@liggitt looks workable to me.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+	// Current version for the protocal interface definition.
+	kmsapiVersion = "v1beta1"
+)
+
+// The gRPC implementation for envelope.Service.
+type gRPCService struct {
+	// gRPC client instance
+	kmsClient  kmsapi.KMSServiceClient
+	connection *grpc.ClientConn
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Why are you keeping this?  It seems like you just need the KMSServiceClient

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	glog.V(4).Infof("KMS provider %s initialized, version: %s", response.RuntimeName, response.RuntimeVersion)
+	return nil
+}
+
+// Decrypt a given data string to obtain the original byte data.
+func (g *gRPCService) Decrypt(cipher []byte) ([]byte, error) {
+	request := &kmsapi.DecryptRequest{Cipher: cipher, Version: kmsapiVersion}
+	response, err := g.kmsClient.Decrypt(context.Background(), request)
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: I would expect timeouts a timeout of timed cancel on the contexts for these calls.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD [hunk 14]
```diff
@@ -0,0 +1,33 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD]: what is generated and what is not generated in this package.  How I regenerate the generated bits?

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go (hunk 9 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go]: this struct is nested inside EncryptionConfig (which has a kind/apiVersion, though not a standard apimachinery-defined one)

the ConfigFile referenced by the KMS can be removed, I think... we have no KMS plugins, and the only method we're planning to support is this external call, right?

**Comment by jdumars** [comment]
@khenidak PTAL

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go (hunk 9 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go]: 1. There will be only gRPC provider, no other plugins. So I will remove "ConfigFile" and the file "plugins.go".
2. There has been "apiVersion" field in config file, if we also add version for KMS, the config file will be:
``` yaml
kind: EncryptionConfig
apiVersion: v1
resources:
  - resources:
    - secrets
    providers:
    - kms:
        apiVersion: v1
        name: grpc
        endpoint: unix:///var/run/kms-provider.sock
```
It seems strange. If there is no other KMS plugins, we can think the "kms" is a part of the whole config.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 10 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: The schema is "unix" not "tcp", so if not provide a dialer, the connection can't be established correctly. We need a way to tell gRPC the protocl is unix domain socket.
Originally we expect support both "unix" and "tcp", now only "unix" socket need to be supported. I will refactor it to make it easier to read.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 12 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: It's for release the connection in test code. Avoid warning message when run test.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 13 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: Okay, I will add it.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD (hunk 14 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD]: Do you mean add scripts similar with "hack/update-generated-runtime.sh" and "hack/update-generated-runtime-dockerized.sh" to generate the "pb.go" file?

**Comment by wu-qiang** [comment]
@deads2k @liggitt Could you please review the latest commit? Thanks.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD (hunk 14 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD]: > Do you mean add scripts similar with "hack/update-generated-runtime.sh" and "hack/update-generated-runtime-dockerized.sh" to generate the "pb.go" file?

Yes.  Or however you created it.  This package looks generated and we need to know how to generate it if we want to update it.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 12 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: > It's for release the connection in test code. Avoid warning message when run test.

Couldn't the KMSServiceClient have a close if you need it?

File: hack/update-generated-kms.sh [hunk 15]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
Comment (deads2k) [inline hack/update-generated-kms.sh]: need a corresponding verify

**Comment by duglin** [comment]
@wu-qiang can you address @deads2k's comments?

**Comment by wu-qiang** [comment]
@duglin I will resolve it ASAP.

**Comment by wu-qiang** [comment]
@duglin @deads2k The comment has been resolved.

**Comment by duglin** [comment]
@wu-qiang thanks for the quick response!

File: hack/update-generated-kms.sh (hunk 15 above)
Comment (duglin) [reply hack/update-generated-kms.sh]: Addressed

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD (hunk 14 above)
Comment (duglin) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD]: @wu-qiang did you address this?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 12 above)
Comment (duglin) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: ?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD (hunk 14 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/BUILD]: yes, "hack/update-generated-kms.sh" and "hack/update-generated-kms-dockerized.sh" are added for it.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go (hunk 12 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go]: KMSServiceClient is gRPC generated code. The field "grpc.ClientConn" is also private, we can't access it outside package.
It's not necessary to expose an extra "Close" method. The gRPC client connection is closed and opened frequently only in test code.

**Comment by duglin** [comment]
@deads2k any other comments?

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 16]
```diff
@@ -56,7 +49,7 @@ resources:
           secret: dGhpcyBpcyBwYXNzd29yZA==
     - kms:
         name: testprovider
-        configfile: testproviderconfig
+        endpoint: /tmp/testprovider.sock
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go]: nit, make the endpoints in the test data actually be valid (`unix:///...`)

**Comment by liggitt** [comment]
no other comments from me. I'd be happy to see a working integration with this (I know the google and vault implementations are in progress)

**Comment by destijl** [comment]
This looks good from our side, @immutableT is just going through the process to get a new repo for the google KMS plugin to live in.

**Comment by khenidak** [comment]
Folks - can we merge this now? @liggitt @destijl

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go (hunk 16 above)
Comment (wu-qiang) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go]: Addressed

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto [hunk 17]
```diff
@@ -0,0 +1,54 @@
+// To regenerate service.pb.go run hack/update-generated-kms.sh
+syntax = "proto3";
+
+package v1beta1;
+
+// This service defines the public APIs for remote KMS provider.
+service KMSService {
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto]: KMSService is redundant. I'd like to fix this this before 1.10. Service names are painful to change.

**Pull request merged**